hash_practice/exercises.py

```diff
@@ -2,28 +2,224 @@
+    for num in nums:
+        if frequency_cells.get(num):
+            frequency_cells[num] += 1
+        else:
+            frequency_cells[num] = 1
+
+    max_frequency = 0
+    key_frequency_cells = {}
+    for num, frequency in frequency_cells.items():
+        if not key_frequency_cells.get(frequency):
+            key_frequency_cells[frequency] = [num]
+        else:
+            key_frequency_cells[frequency].append(num)
+        if frequency > max_frequency:
+            max_frequency = frequency
+
+    result = []
+    while max_frequency > 0 and len(result) < k:
+        if key_frequency_cells.get(max_frequency):
+            for num in key_frequency_cells[max_frequency]:
+                if len(result) < k:
+                    result.append(num)
+        max_frequency -= 1
+
+    return result
+
+
+def top_k_frequent_elements_alternative_solution(nums, k):
+    """ This method will return the k most common elements
+        In the case of a tie it will select the first occuring element.
+
+        pseudocode solution 1:
+        This solution would work for both sorted and not sorted input array.
+        0. Go over the input array and create unique element frequency hash table.
+        1. Go over the frequency hash table and create an array of tuples: [(x1,y1),(x2,y2),(x3,y3),...(xn,yn)]
+        where x1, x2, ..., xn - unique integer elements from the input array
+        and y1, y2, ..., yn their frequencies relatively.
+        2. Sort new array by key, where key is frequency (second element of each tuple).
+        3. Iterate over sorted array k times to form the output by grabbing first element of each tuple and appending it to the resulting array
+
+        Time Complexity: O(nlogn) (sorting)
+        Space Complexity: O(n+k) ?
+
     """
-    pass
+    if len(nums) == k:
+        return nums
+    if len(nums) == 0:
+        return []
+
+    frequency_cells = {}
+    for num in nums:
+        if frequency_cells.get(num):
+            frequency_cells[num] += 1
+        else:
+            frequency_cells[num] = 1
+
+    num_freq_list = []
+    for num, frequency in frequency_cells.items():
+        num_freq_list.append((num, frequency))
+
+    sorted_num_freq_list = sorted(
+        num_freq_list,
+        key=lambda tup: tup[1],
+        reverse=True)
+
+    result = []
+    for i in range(k):
+        result.append(sorted_num_freq_list[i][0])
+
+    return result
 
 
 def valid_sudoku(table):
     """ This method will return the true if the table is still
         a valid sudoku table.
         Each element can either be a ".", or a digit 1-9
-        The same digit cannot appear twice or more in the same 
+        The same digit cannot appear twice or more in the same
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+
+        This code will work for any [n x n] square matrix where n is multiple of 3, thus:
+        Time Complexity: O(n^2)
+        Space Complexity: O(n)
     """

```

Comment:
👍 Technically, the space/time complexity is O(1) since Sudoku boards are always 9x9 and thus will never scale. 

I love the helper functions!